Remotion/Data/DomainObjects.UnitTests/Mapping/TypeDefinitionWalkerWithResultTest.cs

```diff
@@ -0,0 +1,157 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections.Generic;
+using NUnit.Framework;
+using Remotion.Data.DomainObjects.Mapping;
+
+namespace Remotion.Data.DomainObjects.UnitTests.Mapping
+{
+  public class TypeDefinitionWalkerWithResultTest
+  {
+    private class TestableTypeDefinitionWalker<T> : TypeDefinitionWalker<T>
+        where T : class
+    {
+      public Func<ClassDefinition, T> ClassDefinitionHandler { get; set; } = _ => null;
+
+      public Func<InterfaceDefinition, T> InterfaceDefinitionHandler { get; set; } = _ => null;
+
+      public List<TypeDefinition> VisitedTypeDefinitions { get; } = new();
+
+      public TestableTypeDefinitionWalker (TypeDefinitionWalkerDirection direction, Func<T, bool> completionPredicate)
+          : base(direction, completionPredicate)
+      {
+      }
+
+      /// <inheritdoc />
+      public override T VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        VisitedTypeDefinitions.Add(classDefinition);
+
+        var result = ClassDefinitionHandler(classDefinition);
+        if (CompletionPredicate(result))
+          return result;
+
+        return base.VisitClassDefinition(classDefinition);
+      }
+
+      /// <inheritdoc />
+      public override T VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        VisitedTypeDefinitions.Add(interfaceDefinition);
+
+        var result = InterfaceDefinitionHandler(interfaceDefinition);
+        if (CompletionPredicate(result))
+          return result;
+
+        return base.VisitInterfaceDefinition(interfaceDefinition);
+      }
+    }
+
+    [Test]
+    public void VisitClassDefinition_WithChildrenDirectionAndNoFilter_IsCalledInCorrectOrder ()
+    {
+      var (domainBase, person, customer, organizationalUnit) = TypeDefinitionWalkerTest.CreateClassTestDomain();
```

Comment:
this kind of sharing without a base class is a smell. Should we just move this to a base class?